adding alternative TLC59711Multi driver implementation

s-light opened the pull request:
adds TLC59711Multi 
→ this new Implementation is *multi-chip aware*.
the API is compatible with FancyLED (similar to the neopixel/dotstar).

It has a bunch of different options to set pixel values - these have different speed results.
i have add a `examples/tlc59711_multi_dev.py` file with some speed comparisons.
the direct list setter `pixels[pixel_index] = color` supports float and int as tuple / list with auto conversion and extended error raising (reports what goes wrong and why)
this way its a little slower than the *unprotected* alternatives:
the fastest one is `pixels.set_pixel_16bit_value(pixel_index, value_r, value_g, value_b)`

this way the user can use the method that fits his application best..

all tests were done with 128 Pixel (= 32 chips)

pleas let me know what you think about this.

s-light committed:
first starts on multi driver. Not Working.

s-light committed:
fix nameing

s-light committed:
prepared multifile version

s-light committed:
make pylama happy :-)

s-light committed:
make pylama happy :-)

s-light committed:
debugging...

s-light committed:
working :joy:

s-light committed:
fix multichip calculations & remove debugging

s-light committed:
add timing tests

s-light committed:
add lookup variant

s-light committed:
switch to lookup variant

s-light committed:
add set_channel

s-light committed:
optimize set_pixel

s-light committed:
fix spelling

s-light committed:
optimize speed in set_pixel
by inlining functions :-(
faster: 0.16ms per pixel

s-light committed:
optimize set_pixel
by only checking value range once

s-light committed:
optimize set_pixel
more inlinening..

s-light committed:
make linter happy

s-light committed:
optimize same more

s-light committed:
fix spelling

s-light committed:
added set_pixel_16bit_XXXXX

s-light committed:
added timing test

s-light committed:
added test for pixel get

s-light committed:
more tests and cleanup

s-light committed:
add set_pixels_all

s-light committed:
added set_pixel_all_16bit_value

s-light committed:
cleanup & change asserts
to if with helpfull messages

s-light committed:
update

s-light committed:
simplified simpletest.py

s-light committed:
cleaned up *dev.py

s-light committed:
fixed docu and more cleanup of *dev.py

s-light committed:
made pylint complexity checker happy..
this needs about 5ms more @128pixel

s-light committed:
more test & clean up

s-light committed:
add bcdata test sketch

s-light committed:
update and refine BCData things

s-light committed:
some refines for BC-Data

s-light committed:
Merge branch 'master' of https://github.com/adafruit/Adafruit_CircuitPython_TLC59711 into multi

s-light committed:
fix travis for subfolder.

s-light committed:
make sphinx happy ;-)

s-light committed:
fix linting

s-light committed:
more linting fixes

s-light commented:
hm the build fails with `C:  1, 0: Too many lines in module (1136/1000) (too-many-lines)`
but i don't know what to delete...
i know this is a big driver with really a bunch of alternate setters...
and heavily commented...

s-light committed:
and more linting / cleaningup...

s-light committed:
rename internals

for better name compatibility to TLC5957

dhalbert commented on adafruit_tlc59711/adafruit_tlc59711.py:
```diff
@@ -48,21 +48,21 @@
-# access is by design for the internal class.
-#pylint: disable=protected-access
+# context for using internal decorate classes below.
+# In these cases protectected access is by design for the internal class.
```
"protected"

s-light replied:
fixed

dhalbert commented on adafruit_tlc59711/adafruit_tlc59711_multi.py:
```diff
@@ -0,0 +1,1120 @@
+    # helper
+    ##########################################
+
+    CHIP_BUFFER_BYTE_COUNT = 28
```
Do these need to be constants visible to the end user? If not, use `from micropython import const` and then make these `const(...)`, with names that begin with underscores. That will save space on importing.

s-light replied:
done.
i used this for all conts in there. and used the _ internal syntax for all buffer things..
its unlikely that a user needs theme...

dhalbert commented on adafruit_tlc59711/adafruit_tlc59711_multi.py:
```diff
@@ -0,0 +1,1120 @@
+    BUFFER_BYTES_PER_PIXEL = BUFFER_BYTES_PER_COLOR * COLORS_PER_PIXEL
+
+    @staticmethod
+    def set_bit_with_mask(v, mask, x):
```
I don't see this is used anywhere.

dhalbert commented on adafruit_tlc59711/adafruit_tlc59711_multi.py:
```diff
@@ -0,0 +1,1120 @@
+        return v
+
+    @staticmethod
+    def set_bit(v, index, x):
```
I don't see this is used anywhere.

dhalbert commented on adafruit_tlc59711/adafruit_tlc59711_multi.py:
```diff
@@ -0,0 +1,1120 @@
+            value=0
+    ):
+        """Set chip header bits in buffer."""
+        # print(
```
Remove debugging prints

dhalbert commented on adafruit_tlc59711/adafruit_tlc59711_multi.py:
```diff
@@ -0,0 +1,1120 @@
+        header_start = chip_index * self.CHIP_BUFFER_BYTE_COUNT
+        # get chip header
+        header = self._get_32bit_value_from_buffer(header_start)
+        # print("{:032b}".format(header))
```
Remove.

dhalbert commented on adafruit_tlc59711/adafruit_tlc59711_multi.py:
```diff
@@ -0,0 +1,1120 @@
+    ##########################################
+
+    @staticmethod
+    def calculate_Ioclmax(
```
Put args on one line.

dhalbert commented on adafruit_tlc59711/adafruit_tlc59711_multi.py:
```diff
@@ -0,0 +1,1120 @@
+        return Ioclmax
+
+    @staticmethod
+    def calculate_Riref(
```
Put args on one line.

dhalbert commented on adafruit_tlc59711/adafruit_tlc59711_multi.py:
```diff
@@ -0,0 +1,1120 @@
+                "".format(Ioclmax)
+            )
+        if not 0.0 <= IoutR <= Ioclmax:
+            raise ValueError(
```
Put on one or two lines, and below.

s-light replied:
what is your prefered style with this:
my original
```
        if not 2.0 <= Ioclmax <= 60.0:
            raise ValueError(
                "Ioclmax {} not in range: 2mA..60mA"
                "".format(Ioclmax)
            )
```
this is all beautifully aligned in 4spaces..
if i understand you correctly:
```
        if not 2.0 <= Ioclmax <= 60.0:
            raise ValueError("Ioclmax {} not in range: 2mA..60mA"
                "".format(Ioclmax))
```
with this my pylama (pep8) errors with
`E128 continuation line under-indented for visual indent`
so i would need to fi this so:
```
        if not 2.0 <= Ioclmax <= 60.0:
            raise ValueError("Ioclmax {} not in range: 2mA..60mA"
                             "".format(Ioclmax))
```
but this means that the last line is not in multiples of 4 anymore..

please let me know which way i should correct the code :-)
then i will do my best to fix all occurrences..

s-light replied:
have decided to a compromise ;-)
hope that is fine with you. (and i hope i have fixed all occurrences.)

dhalbert commented on adafruit_tlc59711/adafruit_tlc59711_multi.py:
```diff
@@ -0,0 +1,1120 @@
+
+    ##########################################
+
+    def _debug_print_buffer(self):
```
Can remove these debug print helpers now.

dhalbert commented on adafruit_tlc59711/adafruit_tlc59711_multi.py:
```diff
@@ -0,0 +1,1120 @@
+
+    ##########################################
+
+    def _get_32bit_value_from_buffer(self, buffer_start):
```
Use struct instead.

dhalbert replied:
And same for routines below.

s-light replied:
@dhalbert can you give me a hind or link for an example?
i did not know that this is possible in (circuit)python

tannewt replied:
The struct docs for CPython are here: https://docs.python.org/3/library/struct.html We implement most of the same API. You'll want pack_into.

s-light replied:
@tannewt thanks! i forgot about this... 
i implemented it and found that with the struct call in some cases it is very slow.
so after i *cleaned up* all my calls i readded the discreet set in most again for speed reasons.
i have documented this in every function with example times.
(and moved the calls into the functions themselfs - as sub function calls add to much overhead in this case..)

dhalbert commented on adafruit_tlc59711/adafruit_tlc59711_multi.py:
```diff
@@ -0,0 +1,1120 @@
+
+    @staticmethod
+    def _check_and_convert(value):
+        # check if we have float values
```
Could rewrite this as a loop.

s-light replied:
tried this as loop. it is very slow. so i re-changed it to discreet version.
(if the lib is to slow to use with more than 1 chip there is no point in making a multi chip version ;-) )

dhalbert reviewed (changes requested):
Code could be made a lot more compact in various ways. Consider a thorough pass-through to clean up.

s-light commented:
thanks for your review!
i will do the changes and cleanup - but i think i will need some time - to much other things currently going on..

kattni commented:
I'm closing this PR for now. @s-light if you're still interested in submitting this, please either request to reopen this PR or submit a new one from the same branch. Your branch will need to be updated when you're ready to submit it.

s-light commented:
@kattni ok - sorry...
if i get the time i will do it again!

kattni commented:
@s-light No apologies needed! We greatly appreciate you contributing! It's simply that we need to try to keep up with PRs more than we have, and so I'm going through them now. Please feel free to reopen or resubmit when you have time and we'll be happy to review it then!

s-light committed:
tweak

s-light committed:
Merge branch 'master' into multi

s-light committed:
add example

s-light committed:
change to const & internal const

s-light committed:
remove unused and _debug_print_*

s-light committed:
fix code-style (linebreaks)

s-light commented:
@kattni please reopen :-)
iam on the way to fix it!

tannewt commented:
Why do we need a separate driver for more than one IC? Are they chained? Have you seen pixelbuf? It has most of the RGBLED API implemented and C and then delegates to a `_transmit` function for the chip interface.

I'm skeptical we want all of the complexity that this PR introduces. Sorry for not looking at it sooner!

s-light commented:
this drivers are in most use cases chained yes!
and it is 16bit per color/channel - and i tried to design the api as fancyled compatible..
so user don't need to learn some new syntax...

and there is also a recent feature request for just this 'multi-chip' usage in #10.
(that was the motivation to reopen it and do the cleanup... ;-) )


[example use case](https://github.com/s-light/LEDBoard_4x4_16bit) with 4 drivers on one PCB (and the pcbs are also designed to be chained into larger arrangements):
![LEDBoard 4x4 16bit](https://raw.githubusercontent.com/s-light/LEDBoard_4x4_16bit/master/doc/pictures/P1620129_small.jpg)


back when i have started to write the multi version the pixelbuf thing was also just started or at least i did not know of it...
so maybe it would make a lot more sens to write something compatible for this!
→ is it possible to do 16bit per channel values with pixelbuf ?
(otherwise you lose the main feature of this chip.)
as fare as i understand the [documentation](https://circuitpython.readthedocs.io/en/4.x/shared-bindings/_pixelbuf/PixelBuf.html) it is only 8bit.

for all my LEDBoards there is a relative complex arrangement for the led-order - as it is based on possible PCB-routing..
to rearrange this complex oder to a usabel 2D-Array i use a [helper class](https://github.com/s-light/LEDBoard_CP_tester/blob/master/pixel_map.py#L10) for mapping..
maybe this is also something that could be simpliefied / get faster with the pixelbuf thing...
(i don't know what makes more sense here..)


looking forward to your input!

s-light committed:
tweak set&get 16 and 32bit buffer values

tannewt commented:
Thanks for revisiting this!

I don't think pixelbuf will work because it is only 8bits per channel.

Would it be possible to keep this as one module and class that takes in the number of pixels only? That way the same code can run one chip and many.

Having a helper for 2d mapping makes total sense to me. I think @rhooper and @kattni did something similar for NeoPixels.

s-light commented:
i will get back to this end of the week.

> Would it be possible to keep this as one module and class that takes in the number of pixels only? That way the same code can run one chip and many.

this would mean to change the name of the 'multi-class' and tweak the api so that it is only extended (from a current point of view) - so it is backwards compatible.. ?!
i can check what is missing / different and fix this.


currently i copy the helper file from project to project ;-)
if there is interest i can create a repro for it and post it to the community-libraries...

tannewt commented:
> i will get back to this end of the week.
> 
> > Would it be possible to keep this as one module and class that takes in the number of pixels only? That way the same code can run one chip and many.
> 
> this would mean to change the name of the 'multi-class' and tweak the api so that it is only extended (from a current point of view) - so it is backwards compatible.. ?!
> i can check what is missing / different and fix this.

Yup! It'd be great to have just one implementation that works for both.

> 
> currently i copy the helper file from project to project ;-)
> if there is interest i can create a repro for it and post it to the community-libraries...

That'd be great! Thanks!

s-light committed:
fix things..

s-light committed:
more struct experiments and fixes

s-light committed:
linting and code cleanup

s-light committed:
more linting and cleanup

s-light committed:
more cleanup

s-light committed:
rewrite discreet steps as loop

s-light committed:
more cleanup & tests

s-light committed:
cleanup set_channel

s-light committed:
cleanup unneeded helper functions

s-light committed:
speed-optimize

s-light committed:
rename example

s-light committed:
fix bug

s-light committed:
add example

s-light commented:
i have done a bunch of cleanup work and reevaluated some of my code choices.
(documented this and answered in the requested changes by @dhalbert)

regarding the one implementation for both cases:
currently there are some api setters and getters that do not work well for the multichip version  -
or at least would create lots of (time) overhead with - in my opinion - no / less benefits/use-cases:
- the direct `r0`.. `b3` setters
  (if really needed i could map these to the first chip only..)
- the auto_show option
  (if this is something very important for you i try to add it)

i will create a as compatible as possible sub-class of my current multi implementation - 
to try and check what i get done..
(i hope i have it ready end next week..)

tannewt commented:
Thanks for the update @s-light ! Are there any things you'd like me to look at now?

s-light commented:
@tannewt i will let you know if i have don the additions for review. thanks!

kattni commented:
@s-light Are you still working on this? We have made some updates to the library that will require you to update this PR to resolve merge conflicts. Please let us know if you need any assistance with that.

s-light commented:
thanks for letting me know!
yes i am working on this - 
not as constantly as i would like to...

kattni commented:
@s-light No worries! I'm simply checking in. Thanks for replying!

kattni commented:
@s-light Please let me know if you're still working on this. If it will be a while, I'd like to close this PR for now. It can be reopened, and does not affect any of your local code, so all your work is still in place. You can comment on a closed PR and tag me as well in the future to let me know when you're ready to work on it again, if we decide to close it.

s-light commented:
thanks for your frindly reminder ;-)
it is on my list for this week :-)
if its easier for you we can close and reopen if iam ready...

kattni commented:
@s-light We can leave it open, especially if it's on your radar for the near future. Thanks again!

s-light committed:
dev: change debug output

s-light committed:
moved modules outside of subfolder

s-light committed:
changed module names to remove multiple versions

s-light committed:
add backwards compatibility class

s-light committed:
change example import syntax
(not tested yet)

s-light committed:
Merge branch 'master' of https://github.com/adafruit/Adafruit_CircuitPython_TLC59711 into multi

s-light committed:
formating and linting

s-light committed:
run black & linting

s-light commented:
updates :-)
tomorrow i will test all the examples on actual hardware.
i think it is ready. but maybe there needs some tweaks.
currently i don't get the checks to pass.. black formats - but if i check this locally it is all fine..

s-light committed:
cleanup..
(test examples, cleanup doublets, fix small bugs)

s-light committed:
more cleanup
merged TLC59711 into TLC59711Multi & renamed later to first.
if we need a auto_show functionality i think it is necessary to create a 
special subclass for this (for performance reasons)

s-light committed:
added minimal first AutoShow class

s-light committed:
fix examples

s-light committed:
linting..

s-light commented:
now also the AutoShow feature is back- as special extended class..
i have tested all the examples on real hardware so this should be fine now :-)

but i can not get the checks to work -
locally the are fine:
```
/Adafruit_CircuitPython_TLC59711 (multi *+)$ black --check --target-version=py35 .
All done! ✨ 🍰 ✨
9 files would be left unchanged.
```

FoamyGuy commented:
@s-light that is weird. Maybe try running `black --version` to see if you get the same version as actions:
```
black, version 19.10b0
```
I ran locally on your branch and I get same results as the actions: 
```
> black --check --target-version=py35 .
would reformat C:\Users\o_O\Downloads\Adafruit_CircuitPython_TLC59711-multi\examples\tlc59711_fastset.py
would reformat C:\Users\o_O\Downloads\Adafruit_CircuitPython_TLC59711-multi\examples\tlc59711_test_bcdata.py
would reformat C:\Users\o_O\Downloads\Adafruit_CircuitPython_TLC59711-multi\examples\tlc59711_dev.py
would reformat C:\Users\o_O\Downloads\Adafruit_CircuitPython_TLC59711-multi\adafruit_tlc59711.py
Oh no! 💥 💔 💥
4 files would be reformatted, 5 files would be left unchanged.
```

If you have `Allow maintainers to make changes` option enabled I can run black on it here and push the results to your branch if you want in order to try to get make the github CI happy.

s-light commented:
thanks @FoamyGuy 
my version is different (newer):
```
$ black --version
black, version 20.8b1
```
(i freshly installed via pip3 some days ago..)

the `Allow maintainers to make changes` option is enabled.
i am very thankful if you do this!
i am curios what the different versions think should be changed ;-)

FoamyGuy committed:
black format

FoamyGuy committed:
fix pylint issues for super call

s-light commented:
@FoamyGuy thanks for the support!!
i just checked the diff - 

regarding the black version: the [20.8b1 release on github](https://github.com/psf/black/releases) is ~18days old.. 
so maybe will take some time to get to all channels ;-)
and there is a new [magic-comma handling included](https://github.com/mozilla/kitsune/pull/4525#issue-482017646) - 
this explains why the old version truncates *my nice formatting* to one line and the new leaves it in shape ;-)
i would prefere the readable listing - 
and would be willing to use (add)  [`# fmt: on/off`](https://github.com/psf/black/blob/master/docs/the_black_code_style.md#code-style) to fix this for black pre 20.* builds.

thanks for the pylint super) fix!

do you know why the build fails regardless of you fixes?

FoamyGuy commented:
@s-light  The next thing it's upset about is to do with docs building. In the `api.rst` file I think it should have only one entry that looks like this:
```
.. automodule:: adafruit_tlc59711
   :members:
```
I did that on my local copy and it did fix the next actions issue with docs  building however there was another one after that. 

All of the library repos should include a file in the examples directory called `[libraryname]_simpletest.py`. It looks like in your changes you've renamed or removed that file and it's causing the docs build not to work. 

Can you select one of the existing examples that makes a good simple test and rename it to `tlc59711_simpletest.py`. Or if one of the existing ones is not suitable then bring back the old one. I think once that file is back it should get us past the next step. 

Aside from the above: I think at some point we will either A) want to remove one of the files probably the one with `_original` in the name. or B) refactor this into a directory so they can both exist inside of it and user can choose which one to import. 

I don't think I understand the differences between them enough yet to know which would be better though. Is your new version backwards compatible with the old one?

s-light committed:
rename minimal to simpletest

s-light committed:
fix doc

s-light commented:
hmpf - i did not think about / remembered that i have also docu to care about  ;-)...
is `[libraryname]_simpletest.py` a library convention? i did not find a clear *include this files list* in the  [documentation](https://learn.adafruit.com/creating-and-sharing-a-circuitpython-library?view=all)
i personally think the `[libraryname]_minimal.py` name is easier to understand...
and will change it to `[libraryname]_simpletest.py` for now  ;-) ..

the `_original.py` is the original - as i used it to check for api backwards compatibility... 
and no my changes are **not fully backwards compatible** - 
in the new version some things are missing or are renamed.
(to better follow datasheet names / recommendations or create a *nicer* usage interface)

s-light committed:
fixed doc
(removed package sub path)

s-light committed:
another fix

FoamyGuy committed:
black format

FoamyGuy commented:
@s-light I ran Black formatter on it again and we are passing the Actions CI now! I'm not sure if it's possible / how to disable travis CI, but I think we can ignore that one for now.  Thanks for all your work on this.

I'll try to take a deeper dive into this change and test it out but it might take me a week or two in order to get to it.

s-light commented:
@FoamyGuy  thanks for your help on this!!

regarding Travis CI:
[edit]
~~seems like it is expecting a package with the name `adafruit_tlc59711` and tries to lint all subfiles...~~
~~i don't know if this is something that i also screwed up in my configuration...~~
fixed the travis config..[/edit]

take your time.
in a short overview (out of my head): 
the new version is written in wide parts from scratch to control more than one TLC59711 -
so its now a pixel-based approach - not a single-IC based concept..
in this take i also did some timing test and added some fast pixel setter functions.
otherwise it would make no sens to use it at all from CP with lots of pixels...
but i tried and recreate as much of the old api as seemed a good fit to be nearly backwards compatible.
only big difference i currently aware of:
- to have AutoShow activated you have to use a SubClass `TLC59711AutoShow` (for speed reasons)
- BC data api has changed

s-light committed:
fix travis

FoamyGuy committed:
remove travis

FoamyGuy committed:
Merge remote-tracking branch 'origin/multi' into multi

# Conflicts:
#	.travis.yml

FoamyGuy committed:
remove travis

kattni commented:
@s-light Hey! I hope you're well. I wanted to check in to find out where we're at with this. Please let me know when you get a chance!

s-light commented:
@kattni thanks for the reminder!
as far as i know i have finished all open points.
so @FoamyGuy did you had time to check if all is fine from your site?

FoamyGuy commented:
Unfortunately this slipped my mind a bit back after I commented before. My apologies. 

I will try it out this weekend.

FoamyGuy reviewed (changes requested):
I think we are going to have to either get rid of the "original" driver python file, or else convert this into a package (directory) and put both inside of it. As-is I'm pretty sure some parts of the library bundle infrastructure are going to break. My vote would be to get rid of it probably but we can see if Kattni or anyone else has other thoughts.

I tried the new simpletest but I'm not having success with it so far. 
![image](https://user-images.githubusercontent.com/2406189/102017102-678ac300-3d2a-11eb-9f57-5bb74fb94d51.png)
This same circuit does allow me to control LEDs successfully if I revert back to the current master branch driver so I think it's correct, but let me know if I've missed something, I don't have much experience with these chipes. Never used multiples before.

Here is the code I am attempting to use now:
```python
#TLC5971 / TLC59711.

__doc__ = """
tlc59711_minimal.py - TLC59711 minimal usage example.
simple demo of the TLC59711 16-bit 12 channel LED PWM driver.
Shows the minimal usage - how to set pixel values.
Enjoy the colors :-)
"""

import board
import busio

import adafruit_tlc59711

print("tlc59711_minimal.py")

# Define SPI bus connected to chip.
# You only need the clock and MOSI (output) line to use this chip.
spi = busio.SPI(board.SCK, MOSI=board.MOSI)
pixels = adafruit_tlc59711.TLC59711(spi, pixel_count=8)

pixels[0] = (0, 65535, 0)
pixels[4] = (65535, 0, 0)
#pixels[2] = (0.01, 0.0, 0.01)
#pixels[3] = (0.1, 0.01, 0.0)
# if you are ready to show your values you have to call
pixels.show()

# there are a bunch of other advanced ways to set pixel.
# have a look at the other examples.
while True:
    pass

```

I think we should also inlcude a `while True` loop in the simpletest. If I understand correctly, now it exits immediately and may not be apparent that it did anything.

s-light commented:
thanks for testing!!

out of my head i think the problem could be that the order of pixels has changed... :-(
(as these are *shift-register* style data transfer the first send one will be the last pixel in the chain..
and i think i did not compensate for this as most of the time you will need some extra mapping and so this would only add additional delays in the output write....)

i will look into this end of the week..

so if you try 
```
pixels[3] = (100, 65535, 100)
pixels[7] = (65535, 100, 100)
```
does it work?

i thought i already have removed the *_original.py copy...
- the idea was to only have one file and no module directory..
as i tried to be api backwards-compatible as reasonable possible..

kattni commented on .pylintrc:
```diff
@@ -173,9 +173,18 @@ missing-member-max-choices=1
-# List of additional names supposed to be defined in builtins. Remember that
-# you should avoid to define new builtins when possible.
+# List of additional names supposed to be defined in builtins.
+# Remember that you should avoid to define new builtins when possible.
```
I only just noticed this. Under very few circumstances should the `.pylintrc` file be altered. This is not one of those circumstances. Please use `pylint: disable=pylint-failure` within the code if you feel that the errors are not applicable, or try to resolve them if possible. Altering this file makes it difficult for us to update it using a script if we make global changes.

Please revert all of these changes and add `disable`s to your code to handle any issues.

kattni replied:
As a side note - I would prefer the `disable`s and potential `enable`s included not be global. This means I would prefer them to be in-line. However, if you find you're using the same disable repeatedly, it's ok to use a global `disable` but please explain why  if you do. Thanks!

s-light replied:
ok will do this..
i thought its the cleanest way to add the const things..

s-light replied:
done.

kattni reviewed (changes requested):
Thanks for continuing on this! Please see the comment below.

kattni commented:
@s-light With regards to how to structure the library - we can consider a breaking change if it makes sense. There cannot be two .py files at the top level. If the code is getting unwieldy in a single file, or if it seems like the current changes are less likely to be used than the original code, consider creating a package with separate modules. @FoamyGuy can look into what guides etc. would need to be updated and we would try to ensure that happens along with the new release of the library.

FoamyGuy commented:
@s-light the "_original" file might have gotten brought back with a merge or something at some point, I'm not sure. But I do see it in there now in the branch for this PR. I think if you do `git rm adafruit_tlc59711__original.py` and then push a new commit it should take care of that.

I'll try it out with the different indexes tonight. If that is the case I think I am in favor of including a mapping in the library that makes them behave the same as they used to with regards to index orders. Even though most people will end up re-mapping them anyway, I think it's good to have it match the previous order if we can.

FoamyGuy commented:
I tried with the suggested indexes:
```
pixels[3] = (100, 65535, 100)
pixels[7] = (65535, 100, 100)
```
But still not getting any lights on.

s-light commented:
thanks for your feedback!
i will take care of the requests at the weekend!
regarding lib structure: we had this discussion here some time ago and back then was the outcome: only one file.
i also thought i had already removed the _original.py file... 

@FoamyGuy hm that is wired. please try to use the `set_pixel_all((100,10000,1000)` function to see if this works...
i will have to look into this again in detail at the weekend - and will test that the new version indeed works with the original examples..

FoamyGuy commented:
Tried like this:
```python
import board
import busio
import adafruit_tlc59711
print("tlc59711_minimal.py")
spi = busio.SPI(board.SCK, MOSI=board.MOSI)
pixels = adafruit_tlc59711.TLC59711(spi, pixel_count=8)
pixels.set_pixel_all((100,10000,1000))
pixels.show()
print("after show")
while True:
    pass
```
but still not getting any of the LEDs on.

s-light committed:
cleanup

s-light committed:
add detailed explanation

s-light commented:
@FoamyGuy 
i just tested this again with the follwing script:
```
import board
import busio
import adafruit_tlc59711
spi = busio.SPI(board.SCK, MOSI=board.MOSI)
pixels = adafruit_tlc59711.TLC59711(spi)
print('tlc59711 test - light up 4 pixels.')
pixels[0] = (1000, 200, 10)
pixels[1] = (10, 10, 1000)
pixels[2] = (10, 200, 10)
pixels[3] = (500, 10, 1000)
pixels.show()
print('tlc59711 test - show done.')
print('The End')

```

unfortunately i did not find my Adafruit TLC59711 Breakoutboard :-(
so i only can check with my [own made Boards](https://github.com/s-light/LEDBoard_4x4_16bit)
these have four TLC5971 on them.
my boards are fixed setup for 5V Power and Logic - so i have a levelshifter between my ItsyBitsy M4 and the boards.

the most significant difference between the two for us here is the Data-Transfer-Speed:
[TLC5971](https://www.ti.com/product/TLC5971#product-details##features)   20MHz
[TLC59711](https://www.ti.com/product/TLC59711#product-details##features) 10MHz
the protocol itself [is identical](https://e2e.ti.com/support/power-management/f/196/p/493843/1786004#1786004).
in the multi branch i also do no spi configuration other then in the original lib like this:
```
self._spi.configure(baudrate=self._spi.frequency, polarity=0, phase=0)
```
so this should make no difference on the outcome with the given example code.
but maybe the default SPI frequency is different on the different boards?!
but this way with the code the current master branch should also not generate a stabel output...
(i think you used a ESP or similar board in your test-setup?)

for the test i renamed the lib to lib_ so that i am sure there are no other versions used. 
So my CircuitPython Disc looks as this:
```
stefan@stefan-Zen:~$ ls -l /media/stefan/CIRCUITPY/
total 65
-rw-r--r--  1 stefan stefan 39342 Dez 19 12:15 adafruit_tlc59711.py
-rw-r--r--  1 stefan stefan    90 Jan  1  2000 boot_out.txt
drwxr-xr-x 44 stefan stefan 24576 Sep 12 13:28 lib_
-rw-r--r--  1 stefan stefan   337 Dez 19 12:14 main.py
```
my `boot_out.txt`:
```
Adafruit CircuitPython 5.3.0 on 2020-04-29; Adafruit ItsyBitsy M4 Express with samd51g19
```

here are the two tests results as photos:
Original Master:
![P1750872_small](https://user-images.githubusercontent.com/1340319/102692744-b4343980-4215-11eb-814b-b32e16c4d792.jpg)
multi branch
![P1750874_small](https://user-images.githubusercontent.com/1340319/102692747-b72f2a00-4215-11eb-805c-5a21397c0117.jpg)
the first chip on the board is Marked U1 (top left in the image)
(the other chips are showing kind of *garbage* that is passed through...)
and as you can see **indeed the pixel order is changed**.
other than this both work fine.

i try to find my TLC59711 Adafruit Breakout board tomorrow..
(but can not guarantee to find it in my current project chaos... :see_no_evil: )

i have to think about what it would take to correct the pixel order...
(i think i had tested this and it slowed down quite a bit)

s-light committed:
revert changes to .pylintrc
and add file global disables.

(inline disable would not make sens to fix the undefined-variable and 
invalid-name multiple times in my opinion.)

s-light commented:
hmpf the black formating is failing again.
locally it is failing - but for other files ;-)
and this is again the list comma handling...

so are you planing on upgrading the black in the Build CI test?

s-light committed:
add simple fancyled rainbow example

kattni commented:
@s-light 

> so are you planing on upgrading the black in the Build CI test?

Yes. But honestly, only because you brought this to my attention. We pinned `black=19.10b0` and I do not know the reason for this. Just as I pushed for sticking with the latest Pylint to avoid the massive overhaul needed when we moved from 1.9.2 to 2.x, I am going to push for us to move to using the latest Black as well.

My suggestion for now is to force reinstall `black` with the pinned version, and re-run it to see if you can get it working properly. If you need help with that, let me know. I really want to see progress on this PR, and I apologise that CI is causing roadblocks.

s-light committed:
run black==19.10b0

s-light commented:
happy new year to you all!!

all fine.
just as reference for others stumbling over this -
with this line you can downgrade / install a specific black version.
```
pip install black==19.10b0
```
to later upgrade to the latest version you can use
```
pip install --upgrade black
```
([source](https://packaging.python.org/tutorials/installing-packages/))

kattni commented:
@s-light @FoamyGuy Now that this is passing again, where are we at with it? I don't know the whole story so I don't want to jump in and merge if there are still incoming changes.

FoamyGuy commented:
@s-light  this file has `.pdf` extension but is in the examples folder and seems to contain python code. I think it must have gotten pdf accidentally and needs to be changed to `.py`

https://github.com/s-light/Adafruit_CircuitPython_TLC59711/blob/multi/examples/tlc59711_fancyled.pdf

@kattni I was not able to get the examples to run successfully with the new version of the library. Looks like s-light has custom breakouts that may be a bit different than the Adafruit ones perhaps, that is the only guess I have as to the difference in behavior between them and myself. I wonder if it may be possible to get one of the Adafruit breakouts into s-light's hands so they could try these changes out directly on that device.

I've got my circuit still put together stashed away so I can pull it out to re-test any time. But I don't really have any other ideas for troubleshooting at this time.

kattni commented:
@s-light Please email me at kattni@adafruit.com. Thanks!

s-light committed:
fix pdf to py

s-light commented:
the pdf should be py - sorry... (fixed)

i have extended my search for my adafruit breakout board. and luckily i found it :-)
:smile: 

so i have done a second test here.

## testsetup
### hw
- [Adafruit ItsyBitsy M4 Express featuring ATSAMD51](https://www.adafruit.com/product/3800)
- [Adafruit 12-Channel 16-bit PWM LED Driver - SPI Interface - TLC59711](https://www.adafruit.com/product/1455)
- [LED-Board 4x4 16bit](https://github.com/s-light/LEDBoard_4x4_16bit)
(connected in this order)

![P1760016_small](https://user-images.githubusercontent.com/1340319/103892121-4b8cfe00-50eb-11eb-9ad4-931b0710d81e.jpg)
### code i used
```
import board
import busio
import adafruit_tlc59711

spi = busio.SPI(board.SCK, MOSI=board.MOSI)
pixels = adafruit_tlc59711.TLC59711(spi)

print("tlc59711 test - light up 4 pixels.")

pixels[0] = (30000, 200, 200)
pixels[1] = (200, 200, 30000)
pixels[2] = (200, 8000, 200)
pixels[3] = (8000, 200, 8000)
pixels.show()
print("tlc59711 test - show done.")

# these extra show writes shifts the informatoion on to the next chips...
pixels.show()
pixels.show()
pixels.show()
pixels.show()

print("The End")

```

## results

### original
![P1760018_single__crop_small](https://user-images.githubusercontent.com/1340319/103896469-50a17b80-50f2-11eb-8060-1651207346a3.jpg)
### multi
![P1760021_multi__crop_small](https://user-images.githubusercontent.com/1340319/103896480-54350280-50f2-11eb-84e5-c4bfeb7c8d08.jpg)

so for me it is working fine.


## Open Questions
Only thing open (i think):
- do i have to be backwards compatible with the pixel indices order?
  currently it changes (it currently is a raw un-mapped representation of the buffer to write)




---

## Level-Shifting
as i followed [the guide](https://learn.adafruit.com/tlc5947-tlc59711-pwm-led-driver-breakout/python-circuitpython#circuitpython-microcontroller-wiring) for setting up the test- i just questioned myself if i have overlooked something regarding the Digital-Signal Levels and uC with 3.3V logic - 
and checked again in the datasheet:
if you power the Breakout-Board VCC pin (that is connected to the VREG and VCC pin of the chip) with 5V you would have to have also 5V levels for the Signals (SDTI,SCKI) as the [Datasheet states](https://www.ti.com/lit/ds/symlink/tlc59711.pdf#page=4&zoom=250,-123,720):

| PIN | Description | MIN | NOM | MAX | UNIT |
| :-- | :---------- | --: | --: | --: | ---: |
| VREG | Supply voltage, VREG connected to VCC | 3 | 3.3 | 5.5 | V |
| VIH | High-level input voltage (SDTI,SCKI) | 0.7× VREG |  | VREG      | V | 
| VIL | Low-level input voltage (SDTI,SCKI)  | GND       |  | 0.3× VREG | V |

and `0.7V*5V=3.5V` - and that is the minimum for the High-State - so it is not guaranteed to work with 3.3V Signal Levels at 5V VCC, VREF.
so i remembered correctly and to work reliably in the recommend range you need a levelshifter in this configuration.

sadly [the board connects](https://learn.adafruit.com/assets/18728) the TLC59711 VCC to the V+ (+15 in schematics).
if this would be not the case you could use [separate Voltages](https://www.ti.com/lit/ds/symlink/tlc59711.pdf#page=2&zoom=250,-123,562) 
breakoutboard VCC: 3.3V
breakoutboard V+ : 5V
to accomplish this you could ignore the nice V+ pins for the leds and connect them in another way.
i understand that this is a tradeof to not get to complicated for beginner users...

**i just ignored this fact now for the test setup and followed the instructions.**
and of course *it works* ;-)

s-light committed:
fix pylint import order

kattni commented:
@s-light @FoamyGuy Hey! Checking in :). There are some merge conflicts to deal with but they appear to be simple ones.

Where are we at with this?

s-light commented:
i can check on the confilcits next week...
> Where are we at with this?
i think @FoamyGuy did not get it to work at his site at all...
- afteer this i did a test with the adafruit breakout board and it worked for me -
maybe this has to do with the used board... ?!

FoamyGuy commented:
I tested this successfully with two TLC59711 breakouts and can control LEDs connect to both breakouts. Everything seems to be working as intended for me now.

I do think it would be great if we could offer some way to use the index order that matches the silkscreen on the Adafruit breakout, but I don't know if that is a strict requirement. @kattni do you have thoughts on this?

The one other thing I noticed is that the docs don't seem to include the API information any more, and they only have one of the example scripts listed.

Here are the docs built locally from this PR:
![image](https://user-images.githubusercontent.com/2406189/115965068-fdaad300-a4ec-11eb-9e70-08304e483117.png)

and here are the ones live now:
![image](https://user-images.githubusercontent.com/2406189/115965086-14512a00-a4ed-11eb-9ce4-c76c42060a1e.png)

Thanks again for all your work on this @s-light! this is a great new feature to support multiple devices chained together!

kattni commented:
@s-light OK - we are SO close! I was going to resolve the merge conflicts myself through GitHub but the conflicts in adafruit_tlc59711.py are more complicated than I expected. You're the one best suited to resolve them as you know what you added. Is this something you can get fixed up? We can address the documentation in another issue/PR. Once the conflicts are resolved, I want to get this merged. Thank you so much for sticking through this!

@FoamyGuy I don't think it's a strict requirement - I would say file an issue, and consider simply closing it, so the issue is there if anyone searches for it, and someone can make a request there if desired.

jposada202020 commented:
@kattni I could take a look to the documentation once is merged, if that is ok,

kattni commented:
@jposada202020 That would be brilliant! Keep an eye on this PR for when it's merged and then take a look.

jposada202020 commented:
Will do already pin in my inbox :)

s-light committed:
Merge branch 'master' into multi

s-light committed:
merge things cleanup....

s-light commented:
@kattni i think i have resolved all the conflicts - 
and also installed the new *[pre-commit](https://learn.adafruit.com/creating-and-sharing-a-circuitpython-library/check-your-code)* workflow locally... and after sum tweaks it passes ;-)

kattni reviewed (approved):
This is so exciting! Thank you @s-light for all your hard work and perseverance! Thank you @FoamyGuy for testing!

s-light commented:
:smiley: :tada:

jposada202020 commented:
@s-light Thanks :)
